swiftsimio/accelerated.py

```diff
@@ -168,3 +180,148 @@ def index_dataset(handle: Dataset, mask_array: np.array) -> np.array:
     ranges = ranges_from_array(mask_array)
 
     return read_ranges_from_file(handle, ranges, output_size, output_type)
+
+
+################################ ALEXEI: playing around with better read_ranges_from_file implementation #################################
+
+def get_chunk_ranges(ranges, chunk_size, array_length):
+    chunk_ranges = []
+    for bound in ranges:
+        lower = int(np.floor(bound[0]/chunk_size))*chunk_size
+        upper = min(int(np.ceil(bound[1]/chunk_size))*chunk_size, array_length)
+
+        if len(chunk_ranges) > 0:
+            assert(lower >= chunk_ranges[-1][0])
+            assert(upper >= chunk_ranges[-1][1])
+            if lower > chunk_ranges[-1][1]:
+                chunk_ranges.append([lower, upper])
+            elif lower <= chunk_ranges[-1][1]:
+                chunk_ranges[-1][1] = upper
+            else:
+                raise RuntimeError("computing chunk ranges has gone horribly wrong")
+        else:
+            chunk_ranges.append([lower, upper])
+
+    return chunk_ranges
+
+@jit(forceobj=True, fastmath=True)
+def expand_ranges(ranges):
+    length = 0 
+    for bounds in ranges:
+        length += bounds[1] - bounds[0]
+
+    output = np.zeros(length, dtype=np.int64)
+    i = 0
+    for bounds in ranges:
+        lower = bounds[0]
+        upper = bounds[1]
+        bound_length = upper - lower
+        output[i:i+bound_length] = np.arange(lower, upper, dtype=np.int64)
+        i += bound_length
+
+    return output
+
+def extract_ranges_from_chunks(array, chunks, ranges):
+    # Find out which chunk range each range belongs to
+    chunk_array_index = np.zeros(len(ranges), dtype=np.int)
+    chunk_index = 0
+    i = 0
+    while i < len(ranges):
+        if chunks[chunk_index][0] <= ranges[i][0] and chunks[chunk_index][1] >= ranges[i][1]:
```

Comment:
I don't really understand what's going on here - some more comments would be nice and then we can revisit.